### DISCUSSION TITLE
Creating a new Target

### DISCUSSION THREAD

**Discussion by wu2z** [cc65/cc65#2035] [Q&A]
Hello

I am new to the cc65 system. I am very impressed with its power and flexibility.
I was very serious 6502 assembler programmer back in the late 70's and 80's, but havent done any 6502 programming in years. I am not getting back to doing some and need some help.

I am trying to create my own TARGET.  But I don't find any documentation.

I am getting a board that isn't in the system anyplace so I want to create it.  The board I am getting is the UK101.  Then once I have that running I am going to design my own 6502 board so I will need to do it again.

Can any one help me on what to do to create a new target?

Thanks

Keith
ksproul@skychariot.com

**Comment by mrdudz** [comment]
I'd do something like this:
- first get familiar with how you compile and link "manually", that is without using the cl65 frontend. easiest to do this with a target you know and can verify the resulting programs easily (eg C64)
- now learn how to use a custom crt0 and linker config. You can still do this with your familiar target, just copy the existing files, modify them, use them
- at this point you are ready to write some test code for your target. to do this use the "none" target with a custom linker config that defines the memory map for your board. use a custom crt0 that does whatever minimal initialization is needed and then calls _main.
- once this works you can start porting library functions, should you even need them (getting open/read/write/close to work is probably a good idea, as that gives you printf etc)

to add your target to the compiler:
-  add it to the respective list(s) at some places in the code (cant tell the files offhand, usually grep-ing for some other target, like __C64__, gives me what i am looking for :=))
- drop your linker config into the directory with the others (and name it the same as your target)
- make a directory for your target in the libsrc dir, put crt0 and other support files there
- perhaps tackle a makefile or two, i dont remember. most things should just work

(someone should make proper docs for this)

**Comment by wu2z** [comment]
Bob

THANKS!!  This is great and helps a lot.  I still have a lot of learning to do.  The cc65/ca65 System is really fantastic.  I have been reading all of the documentation, but the problem is there aren’t many samples in the docs.

As I said, I used to program when we didn’t have ANYTHING like this.  Now days, I am used to using Make files in ‘C’ on Linux and it works great.

I have Make files working very well with ca65.  I am using Target = NONE and I get exactly what I would expect in the finally binary. Then I wrote a program on Linux to take the file and send it to my eprom P x  programmer (Old GTEK if anyone is curious) and I have exactly what I need.

I tried cloning NONE, but didn’t get very far because I don’t know all of the different files in the different places that have to be duplicated.
I have never programmed on any of the systems that have valid TARGET files, so even trying do them wouldn’t teach me much. And I don’t have any to even try it on.

I have make sort-of working for cc65, it creates a file, but it isn’t loaded correctly, i.e. what I end up with is not an executable file.  So I don’t know all of the steps there.

Now days I do ‘C’ and SQL for Linux and Raspberry Pis, and help my brother with programming huge astronomy projects.

Back in the day, I used the KIM-1, SYM-1, and then the MTU-130 (Micro Technology Unlimited). All 6502 machines.

I am ‘comfortable’ with the Make files, but I am not great with them.
Keith Sproul
Hot Air Balloon Pilot, Special Shapes Pilot
Member BFA AOPA EAA GEBA ARRL NRA
Ham Radio Call Sign WU2Z
***@***.***

**Reply by mrc333777** [reply]
which model gtek do you have? i am trying to get mine updated and proper software etc...

**Comment by polluks2** [comment]
Apropos UK101, we already support the [Ohio Scientific](https://cc65.github.io/doc/osi.html)

**Comment by wu2z** [comment]
Cool!  Can I get the files?  The only things in my target directory are listed below. I don’t see any Ohio entry.

Thanks


apple2
atari
atarixl
c128
c64
cbm610
gamate
geos-cbm
nes
pet
supervision
apple2enh
atari5200
atmos
c16
cbm510
creativision
geos-apple
lynx
pce
plus4
telestrat
vic20

Keith

Keith Sproul
Hot Air Balloon Pilot, Special Shapes Pilot
Member BFA AOPA EAA GEBA ARRL NRA
Ham Radio Call Sign WU2Z
***@***.***

**Comment by polluks2** [comment]
Your list is not up to date, see libsrc/osic1p/
